src/persist-client/src/impl/shard.rs

```diff
@@ -206,6 +206,47 @@ where
         Ok(new_upper)
     }
 
+    pub async fn compare_and_append<'i, I: IntoIterator<Item = ((&'i K, &'i V), &'i T, &'i D)>>(
+        &self,
+        writer_id: &WriterId,
+        updates: I,
+        expected_upper: Antichain<T>,
+        new_upper: Antichain<T>,
+    ) -> Result<Result<Antichain<T>, Antichain<T>>, InvalidUsage> {
+        let mut state = self.state.lock().await;
+
+        match state.writers.get(writer_id) {
+            Some(_) => (), // The writer is known
+            // TODO: This is more likely an expired lease.
+            None => return Err(InvalidUsage(anyhow!("unknown writer: {:?}", writer_id))),
+        };
+
+        let lower = Self::decode_antichain(&state.upper);
```

Comment:
nit: I think i'd find the frontier logic below more intuitive if this was named something like `shard_upper`. feel free to rename it to `lower` before the `for ((k, v), t, d) in updates` loop though